src/main/java/com/coreos/jetcd/internal/impl/Util.java

```diff
@@ -56,4 +68,73 @@ public boolean cancel(boolean mayInterruptIfRunning) {
 
     return targetFuture;
   }
+
+  /**
+   * converts a ListenableFuture of Type S to a CompletableFuture of Type T with retry on
+   * ListenableFuture error.
+   *
+   * @param newSourceFuture a function that returns a new SourceFuture.
+   * @param resultConvert a function that converts Type S to Type T.
+   * @param doRetry a function that determines the retry condition base on SourceFuture error.
+   * @param executor a executor.
+   * @param <S> Source type
+   * @param <T> Converted Type.
+   * @return a CompletableFuture with type T.
+   */
+  static <S, T> CompletableFuture<T> toCompletableFutureWithRetry(
+      Supplier<ListenableFuture<S>> newSourceFuture,
+      Function<S, T> resultConvert,
+      Function<Exception, Boolean> doRetry,
+      Executor executor) {
+
+    AtomicReference<ListenableFuture<S>> sourceFutureRef = new AtomicReference<>();
+    sourceFutureRef.lazySet(newSourceFuture.get());
+
+    CompletableFuture<T> targetFuture = new CompletableFuture<T>() {
+      // the cancel of targetFuture also cancels the sourceFuture.
+      @Override
+      public synchronized boolean cancel(boolean mayInterruptIfRunning) {
+        boolean cancelled = super.cancel(mayInterruptIfRunning);
+        ListenableFuture<S> sourceFuture = sourceFutureRef.get();
+        if (sourceFuture != null) {
+          cancelled = sourceFuture.cancel(true);
+        }
+        return cancelled;
+      }
+    };
+
+    executor.execute(() -> {
+      while (true) {
+        try {
+          ListenableFuture<S> f = sourceFutureRef.get();
+          targetFuture.complete(resultConvert.apply(f.get()));
+          return;
+        } catch (Exception e) {
+          if (doRetry.apply(e)) {
+            synchronized (targetFuture) {
+              if (targetFuture.isCancelled()) {
+                // don't retry if targetFuture has cancelled.
+                return;
+              }
+              sourceFutureRef.set(newSourceFuture.get());
+            }
+            continue;
+          }
+          targetFuture.completeExceptionally(e);
+          return;
+        }
+      }
+    });
+
+    return targetFuture;
+  }
+
+  static boolean isRetriable(Exception e) {
+    Status status = Status.fromThrowable(e);
+    if (status.getCode() == Code.UNKNOWN) {
+      return false;
+    }
+
+    return RETRIABLE_ERRORS.contains(status.getDescription());
```

Comment:
I have another PR that will add better exception handling and it will include exception type checking like you said.